domains/pallets/auto-id/src/lib.rs

```diff
@@ -0,0 +1,519 @@
+                    };
+                    verify_signature(req).ok_or(Error::<T>::InvalidSignature)?;
+                    let validity = Validity::try_from(tbs_certificate.validity)
+                        .map_err(Error::<T>::InvalidValidity)?;
+                    ensure!(
+                        validity.is_valid_at(current_time),
+                        Error::<T>::InvalidValidity(ValidityError::Expired)
+                    );
+
+                    let serial = U256::from_big_endian(&tbs_certificate.serial.to_bytes_be());
+                    issuer_auto_id
+                        .certificate
+                        .issue_certificate_serial::<T>(serial)?;
+
+                    AutoIds::<T>::insert(issuer_id, issuer_auto_id);
+
+                    Certificate::X509(X509Certificate::Leaf(X509CertificateLeaf {
+                        issuer_id,
+                        serial,
+                        subject: tbs_certificate.subject.as_raw().to_vec().into(),
+                        subject_pki: tbs_certificate.subject_pki.raw.to_vec().into(),
+                        validity,
+                        raw: certificate,
+                        revoked: false,
+                    }))
+                }
+            },
+        };
+
+        let auto_id_identifier = NextAutoIdIdentifier::<T>::get();
+        let next_auto_id_identifier = auto_id_identifier
+            .checked_add(Identifier::one())
+            .ok_or(Error::<T>::IdentifierOverflow)?;
+        NextAutoIdIdentifier::<T>::put(next_auto_id_identifier);
+
+        let auto_id = AutoId {
+            identifier: auto_id_identifier,
+            certificate,
+        };
+
+        AutoIds::<T>::insert(auto_id_identifier, auto_id);
+
+        Self::deposit_event(Event::<T>::NewAutoIdRegistered(auto_id_identifier));
+        Ok(())
+    }
+
+    fn do_verify_signature(auto_id: &AutoId, signature: Signature) -> DispatchResult {
+        let Signature {
+            signature_algorithm,
+            value: signature,
+        } = signature;
+        let req = SignatureVerificationRequest {
+            public_key_info: auto_id.certificate.subject_pki(),
+            signature_algorithm,
+            // uses auto_id identifier as the message to sign
+            data: auto_id.identifier.encode(),
+            signature,
+        };
+
+        verify_signature(req).ok_or(Error::<T>::InvalidSignature)?;
+        Ok(())
+    }
+
+    fn do_revoke_certificate(
+        auto_id_identifier: Identifier,
+        signature: Signature,
+    ) -> DispatchResult {
+        let mut auto_id = AutoIds::<T>::get(auto_id_identifier).ok_or(Error::<T>::UnknownIssuer)?;
+        Self::do_verify_signature(&auto_id, signature)?;
+        ensure!(
+            !auto_id.certificate.is_revoked(),
+            Error::<T>::CertificateRevoked
+        );
+
+        auto_id.certificate.revoke();
+        // TODO: revoke all the issued leaf certificates if this is an issuer certificate.
```

Comment:
While that is true, there are potential problems during the verification.
If a root cert is revoked and there are multiple intermediate certificates and in the worst case, if a Leaf certificate owner like prove ownership or something else, then we would have re-check the entire chain of trust onchain and this could be an unbounded operation.

The same goes with revoking where if the root is revoked, then we would have to go through entire list of intermediate certs and revoke them. This is also unbounded operation.

Comparing there two, benefits of the second one is constant verification and verifier does not need to pay for root issuer certificate revocation. Also in the second case, root issuer while revocating the cert will pay the chain of certs revocation.

I'm not completely sure of either and marked a todo to revisit once we have more efficient approaches.